src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -318,4 +319,91 @@ object messages {
         |$code2
         |""".stripMargin
   }
+
+  val implicitClassRestrictionsText =
+    """For a full list of restrictions on implicit classes visit
+      |http://docs.scala-lang.org/overviews/core/implicit-classes.html""".stripMargin
+
+  case class TopLevelImplicitClass(cdef: untpd.TypeDef)(implicit ctx: Context)
+    extends Message(10) {
+    val kind = "Syntax"
+
+    val msg = hl"""|An ${"implicit class"} may not be top-level"""
+
+    val explanation = {
+      val TypeDef(name, impl @ Template(constr0, parents, self, _)) = cdef
+      val exampleArgs = constr0.vparamss(0).map(_.withMods(Modifiers()).show).mkString(", ")
+      val exampleBodyDefs = impl.body.collect {
+        case t if !t.isEmpty => t.show
+      }
+      val exampleBody = if (exampleBodyDefs.isEmpty) ""
+      else exampleBodyDefs.mkString(
+        start = " {\n    ",
+        sep = "\n\n    ",
+        end = "\n  }"
+      )
```

Comment:
I would actually prefer this to have an else clause that is something like ` {\n  ...\n  }`.

My motivation for this is if we have an implicit class with a lot of members, they will all get printed and bloat the message. For copy-pastability (is that even a word?), I think it's fine if we leave out the complete method body.